File: python/lda.py
# -*- coding: utf-8 -*-
"""
Linear Discriminant Analysis (LDA) with Logistic Regression on the Wine Dataset
Created on Thu Jul 3, 2025
@author: USER
"""

# Import necessary libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix
from matplotlib.colors import ListedColormap

# Load dataset
dataset = pd.read_csv('../data/wine.csv')
X = dataset.iloc[:, :-1].values
y = dataset.iloc[:, -1].values

# Split dataset into training and test sets
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=0
)

# Feature scaling
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

# Apply LDA (reduce to 2 linear discriminants)
lda = LinearDiscriminantAnalysis(n_components=2)
X_train = lda.fit_transform(X_train, y_train)
X_test = lda.transform(X_test)

# Train logistic regression model
classifier = LogisticRegression(random_state=0)
classifier.fit(X_train, y_train)

# Predict on test set
y_pred = classifier.predict(X_test)

# Evaluate model
cm = confusion_matrix(y_test, y_pred)
print("Confusion Matrix:\n", cm)

# Function to visualize decision boundary
def plot_decision_boundary(X_set, y_set, title):
    X1, X2 = np.meshgrid(
        np.linspace(X_set[:, 0].min() - 1, X_set[:, 0].max() + 1, 1000),
        np.linspace(X_set[:, 1].min() - 1, X_set[:, 1].max() + 1, 1000)
    )
    Z = classifier.predict(np.c_[X1.ravel(), X2.ravel()]).reshape(X1.shape)

    plt.figure(figsize=(8, 6))
    plt.contourf(X1, X2, Z, alpha=0.25, cmap=ListedColormap(('red', 'green', 'blue')))
    plt.xlim(X1.min(), X1.max())
    plt.ylim(X2.min(), X2.max())

    for index, label in enumerate(np.unique(y_set)):
        plt.scatter(
            X_set[y_set == label, 0],
            X_set[y_set == label, 1],
            c=ListedColormap(('red', 'green', 'blue'))(index),
            label=f'Class {label}'
        )

    plt.xlabel('LD1')
    plt.ylabel('LD2')
    plt.title(title)
    plt.legend()
    plt.grid(True)
    plt.tight_layout()
    plt.show()

# Visualize training set
plot_decision_boundary(X_train, y_train, 'LDA (2D) - Wine Dataset (Training Set)')

# Visualize test set
plot_decision_boundary(X_test, y_test, 'LDA (2D) - Wine Dataset (Test Set)')

File: r/lda.R
# Linear Discriminant Analysis (LDA) with SVM on the Wine Dataset

# Load dataset
dataset <- read.csv('../data/wine.csv')

# Split dataset into training and test sets
library(caTools)
set.seed(123)
split <- sample.split(dataset$Customer_Segment, SplitRatio = 0.8)
dataset_train <- subset(dataset, split == TRUE)
dataset_test  <- subset(dataset, split == FALSE)

# Feature scaling (exclude the label column)
dataset_train[-14] <- scale(dataset_train[-14])
dataset_test[-14]  <- scale(dataset_test[-14])

# Apply LDA for dimensionality reduction
library(MASS)
lda_model <- lda(Customer_Segment ~ ., data = dataset_train)

# Transform datasets using LDA projection
dataset_train <- as.data.frame(predict(lda_model, dataset_train))
dataset_test  <- as.data.frame(predict(lda_model, dataset_test))

# Keep only LD1, LD2, and class label
dataset_train <- dataset_train[c("x.LD1", "x.LD2", "class")]
dataset_test  <- dataset_test[c("x.LD1", "x.LD2", "class")]

# Train SVM classifier
library(e1071)
classifier <- svm(
  formula = class ~ .,
  data = dataset_train,
  type = 'C-classification',
  kernel = 'linear'
)

# Predict on the test set
y_pred <- predict(classifier, newdata = dataset_test[-3])

# Confusion matrix
cm <- table(Actual = dataset_test$class, Predicted = y_pred)
print(cm)

# Function to visualize decision boundary
visualize_results <- function(set, title) {
  X1 <- seq(min(set[,1]) - 1, max(set[,1]) + 1, by = 0.01)
  X2 <- seq(min(set[,2]) - 1, max(set[,2]) + 1, by = 0.01)
  X_grid <- expand.grid(X1, X2)
  colnames(X_grid) <- c("x.LD1", "x.LD2")
  y_grid <- as.numeric(predict(classifier, newdata = X_grid))
  
  plot(set[-3],
       main = title,
       xlab = 'LD1',
       ylab = 'LD2',
       xlim = range(X1),
       ylim = range(X2))
  
  contour(X1, X2, matrix(y_grid, length(X1), length(X2)), add = TRUE)
  points(X_grid, pch = '.', col = ifelse(y_grid == 1, 'tomato',
                                         ifelse(y_grid == 2, 'springgreen', 'skyblue')))
  points(set[-3], pch = 21, col = 'black',
         bg = ifelse(set[,3] == 1, 'darkred',
                     ifelse(set[,3] == 2, 'darkgreen', 'darkblue')))
}

# Visualize results on the training set
visualize_results(dataset_train, 'LDA - Training Set')

# Visualize results on the test set
visualize_results(dataset_test, 'LDA - Test Set')
